lib/Commands/InterruptSignalAwaiter.cpp

```diff
@@ -0,0 +1,75 @@
+//===-- InterruptSignalAwaiter.cpp ----------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2015 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "InterruptSignalAwaiter.h"
+
+#include "llbuild/Basic/PlatformUtility.h"
+
+using namespace llbuild;
+using namespace llbuild::command;
+
+int InterruptSignalAwaiter::signalWatchingPipe[2]{-1, -1};
+std::atomic<bool> InterruptSignalAwaiter::wasInterrupted{false};
+
+InterruptSignalAwaiter InterruptSignalAwaiter::GlobalAwaiter{};
+
+void InterruptSignalAwaiter::sigintHandler(int) {
+  // Set the atomic interrupt flag.
+  wasInterrupted = true;
+
+  // Write to wake up the signal monitoring thread.
+  char byte{};
+  basic::sys::write(signalWatchingPipe[1], &byte, 1);
+}
+
+void InterruptSignalAwaiter::checkForInterrupt() {
+  // Save and clear the interrupt flag, atomically.
+  bool wasInterrupted = InterruptSignalAwaiter::wasInterrupted.exchange(false);
+
+  // Process the interrupt flag, if present.
+  if (wasInterrupted && interruptHandler) {
+    interruptHandler();
+  }
+}
+
+void InterruptSignalAwaiter::signalWaitThread() {
+  // Wait for signal arrival indications.
+  while (true) {
+    char byte;
+    int res = basic::sys::read(signalWatchingPipe[0], &byte, 1);
+
+    // If nothing was read, the pipe has been closed and we should shut down.
+    if (res == 0)
+      break;
+
+    // Otherwise, check if we were awoke because of an interrupt.
+    checkForInterrupt();
+  }
+
+  // Shut down the pipe.
+  basic::sys::close(signalWatchingPipe[0]);
+  signalWatchingPipe[0] = -1;
+}
+
+InterruptSignalAwaiter::InterruptSignalAwaiter() {
+  if (basic::sys::pipe(signalWatchingPipe) < 0) {
+    perror("pipe");
+  }
+
+  new std::thread(&InterruptSignalAwaiter::signalWaitThread, this);
+}
+
+InterruptSignalAwaiter::~InterruptSignalAwaiter() {
+  // Close the signal watching pipe.
+  basic::sys::close(signalWatchingPipe[1]);
+  signalWatchingPipe[1] = -1;
```

Comment:
It doesn't wait for the signal watching thread to complete before finishing the destructor, so the thread associated with the type can outlive it. I was suggesting making the destructor wait for the thread to complete (by using thread::join) before finishing.